README.md
# App para Automatizacion (web scrapping) de Banco Santander
Aplicacion para la gestion de cuentas del banco Santander de Argentina, permite:

* [x] obtener el resumen de la cuenta(download) de los ultimos 60 dias
* [x] transformarlos y cargarlos en una base de datos (sqlite)
* [ ] clasificar y etiquetar los consumos
* [ ] generar reportes

## Instalacion

_se requiere tener instalado python_

Instalar utilizando pip, desde la consola(cmd.exe):

```bash
  pip install my-santander-finance
```

## Actualizacion

Actualizar utilizando pip

```bash
  pip install --upgrade my-santander-finance
```

Luego verificar version
```bash
  sanfi --version
```


## Configuracion

La aplicacion crea un directorio en el 'home' del usuario con el nombre '.sanfi', por ejemplo en Windows seria en:

c:\Users\Oscar\.sanfi\

Para poder realizar el web scrapping de la pagina de Santander Argentina, es necesario definir tres(3) variables de entorno, ya sea como variables de entorno propiamente dichas o bien en un archivo en el raiz del directorio de la app llamado .env, por ejemplo:

c:\Users\Oscar\.sanfi\.env

### Environment Variables

Las tres(3) variables de entorno son:

`DNI`

`CLAVE`

`USUARIO`

_Estos datos, son los requeridos para el login en la web de Santander._ 

Para mas informacion de como trabajar con las variables de entorno hacer click en este link [variables de entorno](docs/es/environment_variables.md)

## Utilizacion

Desde la consola, ejecutar para obtener la ayuda:
```bash
sanfi --help
```

En el caso de querer realizar el download de los consumos:
```bash
sanfi --download
```
La informacion se guarda en una base de datos sqlite (santander.sqlite). Se puede consultar el formato de las tablas en [sqlite](my_santander_finance/sqlite.sql)

Para trabajar directamente con la base de datos sqlite, utilizo la siguiente herramienta grafica free para Windows [HeidiSQL](https://www.heidisql.com/)

Obviamente, tambien es posible utlizar al consola proporcionada por sqlite desde la linea de comandos:

```bash
sqlite3 --help
```
Para mas informacion, click en [sqlite3](docs/es/sqlite3.md)

## Crontab

Para mas informacion, click en [crontab](docs/es/crontab.md)

## Feedback

Contactarme a opaniagu@gmail.com

## Authors

- [@opaniagu](https://www.github.com/opaniagu)


## License

[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
# Changelog

## [Unreleased]

## [0.3.4] - 2023-02-02
### Updated
- Updated chromedriver to release 109

## [0.3.1] - 2022-07-15
### Added
- Fixed typo on import_debito.py


## [0.3.0] - 2022-07-11
### Added
- Add visa credit card web scraping
- Refactor source code

## [0.2.7] - 2022-07-10
### Added
- Add crontab documentation
- Add argument --version
- Fixed on import_debit.py sqlite file path.

## [0.2.6] - 2022-07-10
### Added
- Changed 'app.py' to 'sanfi.py'
- Fixed minor bugs in sanfi.py
- Add log
- Add debug mode
- Add some docs/es files


## [0.2.5] - 2022-07-09
### Added
- Add support automatic download chromedriver on windows (version 103 and 104)

## [0.2.4] - 2022-07-09
### Added
- Add support to home directory on windows
- Add .env.example
- Change app name to "sanfi" (poetry build process - pyproject.toml)


## [0.2.3] - 2022-07-08
### Added
- Add support to lint


## [0.2.2] - 2022-07-08
### Added
- Rename downloaded files to debit_yyyy-mm-dd_hh#mm#ss.xls mask.


## [0.2.1] - 2022-07-07
### Added
- Validate (empty) env or .env variables, before start to login to bank url.


## [0.2.0] - 2022-07-07
### Added
- Start using "changelog" format based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
- Start using "version" based on [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
- Add control version using bump
- Add Makefile
- Add docs folder with language support (en, es)

### Fixed
- Fix Settings CHROME_DRIVER_DIR to find "chromedriver.exe" based on ROOT_DIR


### Changed
- Start using 'en' (English) as default language comment code


my_santander_finance/init.py
import os
import shutil
from os.path import join

import requests
import sqlalchemy

from my_santander_finance.config.settings import settings
from my_santander_finance.logger import Logger
from my_santander_finance.util.func_chrome import get_chrome_version
from my_santander_finance.util.func_dir import create_dir

log = Logger().get_logger(__name__)


def init_dir():
    create_dir(settings.LOCAL_DIR)
    create_dir(settings.DOWNLOAD_DIR)
    create_dir(settings.CVS_TEMP_DIR)
    create_dir(join(settings.LOCAL_DIR, "driver"))

    # debit
    create_dir(settings.DOWNLOAD_CUENTA_DIR)
    create_dir(settings.DOWNLOAD_CUENTA_DIR + ".old")

    # visa
    create_dir(settings.DOWNLOAD_VISA_DIR)
    create_dir(settings.DOWNLOAD_VISA_DIR + ".old")

    # amex
    create_dir(settings.DOWNLOAD_AMEX_DIR)
    create_dir(settings.DOWNLOAD_AMEX_DIR + ".old")


def sqlite_exec_sql(sql: str):
    # conectar a la base de datos
    SQLALCHEMY_DATABASE_URI = "sqlite:///" + os.path.join(settings.LOCAL_DIR, settings.DATABASE_SQLITE)
    log.debug(f"sqlite uri '{SQLALCHEMY_DATABASE_URI}'")
    database_connection = sqlalchemy.create_engine(SQLALCHEMY_DATABASE_URI, echo=False)

    # ejecutar la consulta sql
    try:
        conn = database_connection.connect()
        conn.execute(sql)
        # database_connection.execute(sql)

    except sqlalchemy.exc.SQLAlchemyError as ex:
        # Silently ignore errors if table and index already exist
        if str(ex).find("already exists") != -1:
            log.debug("sql skiped...resource already exist")
        else:
            # error = str(ex.__dict__["orig"])
            # log.debug(error)
            log.debug(ex)


def init_sqlite():
    # debito
    sql = """
            CREATE TABLE "debit" (
                "id" INTEGER PRIMARY KEY AUTOINCREMENT,
                "fecha" DATE NULL,
                "sucursal_origen" TEXT NULL,
                "descripcion" TEXT NULL,
                "referencia" BIGINT NULL,
                "cuenta_sueldo" REAL DEFAULT 0.0,
                "importe_cuenta_corriente_pesos" REAL DEFAULT 0.0,
                "saldo_pesos" REAL DEFAULT 0.0,
                "tarjeta" TEXT NULL,
                "categoria" TEXT NULL,
                "nota" TEXT NULL
        )
        ;
        """
    sqlite_exec_sql(sql)

    sql = """
            CREATE UNIQUE INDEX `index_1` ON debit (`fecha`, `descripcion`,`cuenta_sueldo`);
          """
    sqlite_exec_sql(sql)

    # visa
    sql = """
            CREATE TABLE "visa" (
                "id" INTEGER PRIMARY KEY AUTOINCREMENT,
                "fecha" DATE NULL,
                "descripcion" TEXT NULL,
                "establecimiento" TEXT NULL,
                "comprobante" TEXT NULL,
                "importe_pesos" REAL DEFAULT 0.0,
                "importe_dolares" REAL DEFAULT 0.0,
                "tarjeta" TEXT NULL,
                "categoria" TEXT NULL,
                "nota" TEXT NULL
        )
        ;
        """
    sqlite_exec_sql(sql)

    sql = """
            CREATE UNIQUE INDEX `index_2` ON visa (`fecha`, `descripcion`,`comprobante`);
          """
    sqlite_exec_sql(sql)

    # amex
    sql = """
            CREATE TABLE "amex" (
                "id" INTEGER PRIMARY KEY AUTOINCREMENT,
                "fecha" DATE NULL,
                "descripcion" TEXT NULL,
                "establecimiento" TEXT NULL,
                "comprobante" TEXT NULL,
                "importe_pesos" REAL DEFAULT 0.0,
                "importe_dolares" REAL DEFAULT 0.0,
                "tarjeta" TEXT NULL,
                "categoria" TEXT NULL,
                "nota" TEXT NULL
        )
        ;
        """
    sqlite_exec_sql(sql)

    sql = """
            CREATE UNIQUE INDEX `index_3` ON amex (`fecha`, `descripcion`,`comprobante`);
          """
    sqlite_exec_sql(sql)


def create_env_example():
    env_file_path = join(settings.LOCAL_DIR, ".env.example")
    if os.path.exists(env_file_path) is False:
        with open(env_file_path, "w") as f:
            f.write("DNI=12345678\n")
            f.write("CLAVE=clave\n")
            f.write("USUARIO=usuario\n")


def download_chromedriver():
    """only supported windows OS, chrome"""
    if os.path.exists(settings.CHROME_DRIVER_EXE) is False:
        chrome_version = get_chrome_version()[:3]
        if chrome_version == "103":
            url = "https://chromedriver.storage.googleapis.com/103.0.5060.53/chromedriver_win32.zip"
        elif chrome_version == "104":
            url = "https://chromedriver.storage.googleapis.com/104.0.5112.29/chromedriver_win32.zip"
        elif chrome_version == "109":
            url = "https://chromedriver.storage.googleapis.com/109.0.5414.74/chromedriver_win32.zip"
        elif chrome_version == "110":
            url = "https://chromedriver.storage.googleapis.com/110.0.5481.30/chromedriver_win32.zip"
        else:
            print(f"Can not found chromedriver version {chrome_version}")
            return

        # try to download
        r = requests.get(url, allow_redirects=True)
        open(settings.CHROME_DRIVER_ZIP, "wb").write(r.content)
        shutil.unpack_archive(settings.CHROME_DRIVER_ZIP, settings.CHROME_DRIVER_DIR)


# ----------------------------------
if __name__ == "__main__":
    init_dir()
    create_env_example()
    init_sqlite()
    download_chromedriver()
